jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PerTest.scala

```diff
@@ -107,4 +108,34 @@ trait PerTest extends BeforeAndAfterEach { self: FunSuite =>
     ss.createDataFrame(sc.parallelize(it.toList, numPartitions))
       .toDF("id", "label", "features", "group")
   }
+
+  // Background:
+  // 1. XGBoost has enabled Stopping SparkContext by default while tasks run into exceptions,
+  // 2. The Stopping SparkContext (SparkContext.getOrCreate().stop()) is called in a separate
+  //    thread differring with thread running unit tests
+  // 3. The unit tests are executed one by one in a same thread.
+  //
+  // Consider this sitution, TEST B follows TEST A (TEST A will throw exception in the worker)
+  // a) TEST A will trigger SparkContext.getOrCreate().stop() which takes long time to be finished
+  //        in a separate thread
+  // b) TEST A calls afterEach which tries to stop SparkContext which is stopping because of a)
+  // c) TEST B runs. first, calling beforeEach which creates SparkSession by wrapping a SparkContext
+  //      called by SparkContext.getOrCreate(), Since SparkContext is a singleTon and kept by
+  //      activeContext which was defined in SparkContext,
+  //
+  // SparkContext.stop will clear activeContext in the last of stop by
+  //      SparkContext.clearActiveContext()
+  // Here is the issue,
+  // The newly created Sparksession of TEST B will wrap the SparkContext which is stopping by
+  // TEST A, which will result sparkContext.assertNotStopped() throw exception and block the
+  //  following unit tests in SparkSession.
+
+  private def waitSparkContextTotallyStopped: Unit = {
+    var totalWaitedTime = 0L
+    while (!SparkContextUtils.getActiveSparkContext.isEmpty && totalWaitedTime <= 11000) {
+      Thread.sleep(1000)
```

Comment:
Is this really a robust solution?

Reply:
Yes, it's not a robust solution for now, since we really don't know when the SparkContext will be stopped in other thread. What we can do is set a reasonable timeout and check. I just checked the code, seems there is an explicit 21s timeout in the "stop" code. maybe I need to change the timeout to 22s.